media/Mondrian/mondrianViewer.js

```diff
@@ -0,0 +1,265 @@
+
+  if (totalMemory === undefined) {
+    totalMemory = 0;
+  }
+
+  segmentSelect.replaceChildren();
+  for (const [index, segment] of data.segments.entries()) {
+    segmentSelect.appendChild(new Option(segment.name, index));
+  }
+  segmentSelect.value = viewer.activeSegment;
+
+  for (alloc of data.segments[viewer.activeSegment].allocations) {
+    if (alloc.alive_till > totalCycles) {
+      totalCycles = alloc.alive_till;
+    }
+
+    if (alloc.offset + alloc.size > totalMemory) {
+      totalMemory = alloc.offset + alloc.size;
+    }
+  }
+
+  memorySizeContainer.innerText = `${totalMemory}`;
+  cycleCountContainer.innerText = `${totalCycles}`;
+
+  let loadMs = (performance.now() - loadTs).toFixed(2);
+  statusLineContainer.innerText = `Document loaded in ${loadMs}ms`;
+
+  viewer.viewportMinCycle = 0;
+  viewer.viewportMaxCycle = totalCycles;
+  viewportMemory = totalMemory;
+
+  updateViewport(data, viewer);
+}
+
+function scaleViewport(viewer) {
+  const scrollContainer = viewerContainer.parentElement;
+  const scrollH = scrollContainer.scrollLeft / scrollContainer.scrollWidth;
+  const scrollV = scrollContainer.scrollTop / scrollContainer.scrollHeight;
+
+  const scaleH = Math.pow(2, viewer.viewportHScale);
+  const scaleV = Math.pow(2, viewer.viewportVScale);
+
+  const viewportCycles = viewer.viewportMaxCycle - viewer.viewportMinCycle;
+  viewerContainer.style.width = viewportCycles * scaleH + 'px';
+  viewerContainer.style.height = viewportMemory * scaleV / 8192 + 'px';
+  scrollContainer.scrollLeft = scrollContainer.scrollWidth * scrollH;
+  scrollContainer.scrollTop = scrollContainer.scrollHeight * scrollV;
+
+  if (scaleH < 4 || scaleV < 4) {
+    viewerContainer.style.backgroundImage = 'none';
+  } else {
+    viewerContainer.style.backgroundImage = `url("data:image/svg+xml;charset=UTF-8,%3csvg ` +
+        `xmlns='http://www.w3.org/2000/svg' width='${scaleH}' height='${scaleV}'%3e%3cpath ` +
+        `style='fill:none;stroke-width:1px;stroke:%23fff;opacity:0.1' ` +
+        `d='M 0,${scaleV - 0.5} ${scaleH - 0.5},${scaleV - 0.5} ${scaleH - 0.5},0' ` +
+        `/%3e%3c/svg%3e")`;
+  }
+
+  if (viewer.viewportHScale < 3) {
+    viewerContainer.classList.add('mondrian-viewer-bounds-no-label');
+  } else {
+    viewerContainer.classList.remove('mondrian-viewer-bounds-no-label');
+  }
+}
+
+function updateViewport(data, viewer) {
+  let boxTemplate = document.createElement('div');
+  boxTemplate.classList.add('mondrian-allocation-box');
+
+  let boxTemplateLabel = document.createElement('div');
+  boxTemplateLabel.classList.add('mondrian-allocation-label');
+
+  boxTemplate.appendChild(boxTemplateLabel);
+
+  viewerContainer.replaceChildren();
+  scaleViewport(viewer);
+
+  const viewportCycles = viewer.viewportMaxCycle - viewer.viewportMinCycle;
+  for (const [i, alloc] of data.segments[viewer.activeSegment].allocations.entries()) {
+    if (alloc.alive_from > viewer.viewportMaxCycle) {
```

Comment:
As I am not good at javascript,  please understand me even I ask some stupid questions :)
( I am glad that you created a good example with custom editor because I am working with custom editor too :D )

---

From line 131, following codes can be found.

```js
  for (alloc of data.segments[viewer.activeSegment].allocations) {
    if (alloc.alive_till > totalCycles) {
      totalCycles = alloc.alive_till;
    }
    // ...
viewer.viewportMaxCycle = totalCycles;
```

Then... is this case can be occurred?

Reply:
Oh... this check was added in advance before implementation of the next feature - the ability to show only a part of the model (large models with 100k+ cycles can slow things down quite a bit... that's the purpose of `viewportMinCycle` / `viewportMaxCycle` properties - this check skips allocations outside of selected range). I haven't implemented it yet (I need to implement interactive range selector first), so I plan to add this feature in the next PR :)